### PR TITLE
[3] Implement the negation logic for predefined and full-text queries

### PR DESCRIPTION
**Pull request by KyleJu**
Fix https://github.com/GoogleChrome/chromium-dashboard/issues/2347. Preprocess the negation queries so that it doesn't only affect the term that immediately before it.
- Update the logical operator parser for both structured and full-text queries, to parse `OR ` (with trailing space) and '-' (without trailing space)
- Turn all negation operations into a AND operation

### CODE REVIEW

File: internals/search.py [hunk 1]
```diff
@@ -279,3 +287,34 @@ def process_query(
...
+
+    if negated_id_set is None:
+        negated_id_set = fetch_all_feature_ids_set()
+
+    feature_ids = _resolve_promise_to_id_list(future)
+    negated_id_set.difference_update(feature_ids)
+
+  # Append the result of negation operations to the end.
+  if negated_id_set is not None:
+    new_future_ops.append(('', list(negated_id_set)))
```
Comment (jrobbins) [inline internals/search.py:312]: Could we append `('', negated_id_set)` without converting to a list?

File: internals/search_test.py [hunk 2]
```diff
@@ -55,14 +55,14 @@ def test_structured_query_terms__complex(self):
         [('-', 'field', '=', 'value', '')],
         search.TERM_RE.findall('-field=value '))
     self.assertEqual(
-        [('-', 'field', '>', 'value', '')],
-        search.TERM_RE.findall('-field>value '))
+        [('OR ', 'field', '>', 'value', '')],
```
Comment (jrobbins) [inline internals/search_test.py:58]: You have some `OR` unit test cases mixed into this negation PR.

**Review by jrobbins — APPROVED** [review]
LGTM after considering a couple of comments.

File: internals/search.py (hunk 1 above)
Comment (KyleJu) [reply internals/search.py:312]: Yup we could!

File: internals/search_test.py (hunk 2 above)
Comment (KyleJu) [reply internals/search_test.py:58]: I also added `OR ` parsing in this PR.

File: internals/search.py (hunk 1 above)
Comment (KyleJu) [reply internals/search.py]: I got too many complaints from mypy. I will change it in the next PR, along with the OR logic

**Pull request merged**